benchmark/benchmark.hpp

```diff
@@ -441,20 +448,164 @@ class multiple_malloc_free_benchmark : public benchmark_interface<Size, Alloc> {
         }
     }
 
-    void prepareWorkload(benchmark::State &state) {
+    virtual void prepareWorkload(benchmark::State &state) {
         auto tid = state.thread_index();
         auto &n = next[tid];
+
+        // Create generators for random index selection and binary decision.
+        using distribution = std::uniform_int_distribution<size_t>;
         std::default_random_engine generator;
-        distribution dist;
+        distribution dist_offset(0, max_allocs - 1);
+        distribution dist_opt_type(0, 1);
         generator.seed(0);
-        dist.param(distribution::param_type(0, max_allocs - 1));
+
         auto sizeGenerator = base::alloc_sizes[tid];
+        std::vector<size_t> free;
+        std::vector<size_t> allocated;
+
+        // Preallocate memory: initially, half the indices are allocated.
+        // See prealloc() function;
+        size_t i = 0;
+        for (; i < max_allocs / 2; i++) {
+            allocated.push_back(i);
```

Comment:
Reserve memory for `free` and `allocated` vectors for efficiency. The required for this initialization sizes are already known.

Reply:
Additionally twice:
```cpp
while (i < max_allocs / 2) {
    allocated.push_back(i++);
}
```

Reply:
done